master/buildbot/reporters/telegram.py

```diff
@@ -0,0 +1,1055 @@
+        if args.lower() == 'botfather':
+            lp = len(self.bot.commandPrefix)
+            commands = self.build_commands()
+            response = []
+            for command in commands:
+                command = command[lp:]
+                if command == 'start':
+                    continue
+                meth = self.getCommandMethod(command, True)
+                doc = getattr(meth, '__doc__', None)
+                if not doc:
+                    doc = command
+                response.append("{} - {}".format(command, doc))
+            if response:
+                self.send('\n'.join(response))
+        else:
+            return super().command_COMMANDS(args)
+
+    @defer.inlineCallbacks
+    @Contact.overrideCommand
+    def command_DANCE(self, args, **kwargs):
+        chat = self.channel['id']
+        msg = yield self.send("**<(^.^<)**")
+        if msg is not None:
+            mid = msg['message_id']
+            self.bot.reactor.callLater(1.0, self.bot.edit_message, chat, mid, "**<(^.^)>**")
+            self.bot.reactor.callLater(2.0, self.bot.edit_message, chat, mid, "**(>^.^)>**")
+            self.bot.reactor.callLater(2.5, self.bot.edit_message, chat, mid, "**(7^.^)7**")
+            self.bot.reactor.callLater(4.0, self.bot.edit_message, chat, mid, "**(>^.^<)**")
+            self.bot.reactor.callLater(5.0, self.bot.delete_message, chat, mid)
+            self.bot.reactor.callLater(5.5, self.bot.send_sticker, chat, random.choice((
+                'CAADAgAD9wEAAsoDBgtCnbBFfI8M_BYE',
+                'CAADAgADQQIAArnzlwuD160COMLwKRYE')))
+
+    @defer.inlineCallbacks
+    def command_GETID(self, args, **kwargs):
+        """get user and chat ID"""
+        if self.userid == self.chatid:
+            self.send("Your ID is {}.".format(self.userid))
+        else:
+            yield self.send("{}, your ID is {}.".format(self.user_name, self.userid))
+            self.send("This {} ID is {}.".format(self.channel.get('type', "group"), self.chatid))
+    command_GETID.usage = "getid - get user and chat ID that can be put in the master configuration file"
+
+    @defer.inlineCallbacks
+    @Contact.overrideCommand
+    def command_LIST(self, args, **kwargs):
+        args = self.splitArgs(args)
+        if not args:
+            keyboard = [
+                [self.query_button("👷️ Builders", '/list builders')],
+                [self.query_button("⚙ Workers", '/list workers')],
+                [self.query_button("📄 Changes", '/list changes')],
+            ]
+            self.send("What do you want to list?",
+                      reply_markup={'inline_keyboard': keyboard})
+            return
+
+        all = False
+        num = 5
+        try:
+            num = int(args[0])
+            del args[0]
+        except ValueError:
+            if args[0] == 'all':
+                all = True
+                del args[0]
+        except IndexError:
+            pass
+
+        if all:
+            num = 20
```

Comment:
Why are we limiting the amount of returned data if the user explicitly requested `all`?

Reply:
Because of the Telegram message length limit. Actually I had to limit it to 10. I have added a comment, so the user knows that the number of shown changes has been limited.

Reply:
Would it be possible to show multiple messages if a single message is overflown?

Reply:
Done. However, listing all changes can be problematic as there can be thousands of them. I have added additional confirmation.

Reply:
Sounds great, thanks!